Real-time message processing and synchronized database updates: plugin support

nateyazdani opened the pull request:
This brings the plugins in-line with the new architecture proposed by the [associated pull request to statscache](https://github.com/fedora-infra/statscache/pull/28), and they should be considered jointly.

nateyazdani committed:
Split handle() method into process() and update()

Disassociate message processing and model updating, in accordance with the
plugin system restructuring instituted by the related commit in statscache
proper ("Restructure plugin system").

nateyazdani committed:
Eliminate initializa() methods

Remove the initialize() method from all plugins, as that functionality will
now be performed by the stats consumer on initialization (as described in the
commit "Centralize backlog query" in the main statscache repository).

nateyazdani committed:
Explicitly give dummy epochs for dummy Frequencies

Necessary, as the epoch parameter is now mandatory.

nateyazdani committed:
Rename Frequency to Schedule

Necessary to support the renaming of the class in statscache proper (there is an
accompanying commit by the same headline).

nateyazdani committed:
Remove unnecessary variables

ralphbean commented on statscache_plugins/volume/by_category.py:
```diff
@@ -14,28 +13,31 @@ class PluginMixin(VolumePluginMixin):
+        self._volumes[(category, timestamp)] += 1
 
-        for key, volume in volumes.items():
+    def update(self, session):
```
It may be that these `update` methods can be simplified even more.

Lots of their logic is about detecting previous values stored at the 'current' timestamp by some previous partial run, and then incrementing that stored value.

Now, plugins are keeping their running queue (`self._volumes` in this case) in memory, and then only flushing it out to disk later.

We had previously off-loaded this in-memory state to the DB to be able to recover from unexpected crashes, but it seems that the central statscache bits are now responsible for figuring out the missed workload on behalf of plugins, freeing the plugins themselves of this responsibility.

nateyazdani replied:
Very true, these methods are boilerplate code to a large extent (or at least, the volume ones are). Yep, your understanding of that is correct; plugins need only know how to process a message and update the model. I'll play around with this and push a change to consolidate the common logic into the `VolumePluginMixin` class, most likely.

nateyazdani committed:
Refactor model update logic for volume plugins

Due to the similarities in the update() method of the volume plugin, the
behavior has been generalized and moved into the main VolumePluginMixin class.

ralphbean commented:
Nice!  Looks good.  :+1: here.

``` python
func(**dict(zip(a, b))) # ftw!
```

nateyazdani commented:
Awesome, merging :)
